chat-android/src/main/java/com/ably/chat/Logger.kt

```diff
@@ -0,0 +1,141 @@
+    tag: String? = null,
+    staticContext: Map<String, String> = mapOf(),
+    dynamicContext: Map<String, () -> String> = mapOf(),
+): LogContext {
+    return LogContext(
+        tag = tag ?: this.tag,
+        staticContext = this.staticContext + staticContext,
+        dynamicContext = this.dynamicContext + dynamicContext,
+    )
+}
+
+internal class AndroidLogger(
+    private val minimalVisibleLogLevel: LogLevel,
+    override val context: LogContext,
+) : Logger {
+
+    override fun withContext(tag: String?, staticContext: Map<String, String>, dynamicContext: Map<String, () -> String>): Logger {
+        return AndroidLogger(
+            minimalVisibleLogLevel = minimalVisibleLogLevel,
+            context = context.mergeWith(tag, staticContext, dynamicContext),
+        )
+    }
+
+    override fun log(message: String, level: LogLevel, throwable: Throwable?, newTag: String?, newStaticContext: Map<String, String>) {
+        if (level.logLevelValue <= minimalVisibleLogLevel.logLevelValue) return
+        val finalContext = context.mergeWith(newTag, newStaticContext)
+        val tag = finalContext.tag
+        val completeContext = finalContext.staticContext + finalContext.dynamicContext.mapValues { it.value() }
+
+        val contextString = ", context: $completeContext"
+        val formattedMessage = "[${LocalDateTime.now()}] $tag ${level.name} ably-chat: ${message}$contextString"
+        when (level) {
+            // We use Logcat's info level for Trace and Debug
+            LogLevel.Trace -> Log.i(tag, formattedMessage, throwable)
+            LogLevel.Debug -> Log.i(tag, formattedMessage, throwable)
+            LogLevel.Info -> Log.i(tag, formattedMessage, throwable)
+            LogLevel.Warn -> Log.w(tag, formattedMessage, throwable)
+            LogLevel.Error -> Log.e(tag, formattedMessage, throwable)
+            LogLevel.Silent -> {}
+        }
+    }
+}
+
+internal class CustomLogger(
+    private val logHandler: LogHandler,
+    private val minimalVisibleLogLevel: LogLevel,
+    override val context: LogContext,
+) : Logger {
+
+    override fun withContext(tag: String?, staticContext: Map<String, String>, dynamicContext: Map<String, () -> String>): Logger {
+        return CustomLogger(
+            logHandler = logHandler,
+            minimalVisibleLogLevel = minimalVisibleLogLevel,
+            context = context.mergeWith(tag, staticContext, dynamicContext),
+        )
+    }
+
+    override fun log(message: String, level: LogLevel, throwable: Throwable?, newTag: String?, newStaticContext: Map<String, String>) {
+        if (level.logLevelValue <= minimalVisibleLogLevel.logLevelValue) return
+        val finalContext = context.mergeWith(newTag, newStaticContext)
+        logHandler.log(
+            message = message,
+            level = level,
+            throwable = throwable,
+            context = finalContext,
+        )
+    }
+}
+
+internal class EmptyLogger(override val context: LogContext) : Logger {
```

Comment:
Is use-case of `EmptyLogger` is limited to tests, then it can be moved to `TestUtils.kt` and rename it to `TestLogger` wdyt